File: TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/map/IconUtils.java
package com.skedgo.tripkit.ui.map;

import android.content.res.Resources;
import android.util.DisplayMetrics;

public final class IconUtils {
  private IconUtils() {}

  public static String asUrl(Resources resources, String icon, String urlTemplate) {
    final String densityDpiName = getDensityDpiName(resources.getDisplayMetrics().densityDpi);
    return String.format(urlTemplate, densityDpiName, icon);
  }

  private static String getDensityDpiName(int densityDpi) {
    switch (densityDpi) {
      case DisplayMetrics.DENSITY_MEDIUM:
        return "mdpi";
      case DisplayMetrics.DENSITY_HIGH:
        return "hdpi";
      case DisplayMetrics.DENSITY_XHIGH:
        return "xhdpi";
      case DisplayMetrics.DENSITY_XXHIGH:
        return "xxhdpi";
      case DisplayMetrics.DENSITY_XXXHIGH:
      default:
        return "xxxhdpi";
    }
  }
}

File: TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/map/IconUtils.kt
package com.skedgo.tripkit.ui.map

import android.content.res.Resources
import android.util.DisplayMetrics
import androidx.annotation.VisibleForTesting

object IconUtils {
    fun asUrl(resources: Resources, icon: String, urlTemplate: String): String {
        val densityDpiName = getDensityDpiName(resources.displayMetrics.densityDpi)
        return String.format(urlTemplate, densityDpiName, icon)
    }

    @VisibleForTesting
    fun getDensityDpiName(densityDpi: Int): String {
        return when (densityDpi) {
            DisplayMetrics.DENSITY_MEDIUM -> "mdpi"
            DisplayMetrics.DENSITY_HIGH -> "hdpi"
            DisplayMetrics.DENSITY_XHIGH -> "xhdpi"
            DisplayMetrics.DENSITY_XXHIGH -> "xxhdpi"
            else -> "xxxhdpi"
        }
    }
}
